samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java

```diff
@@ -77,6 +80,7 @@ public SamzaContainerController(
     } else {
       this.containerShutdownMs = containerShutdownMs;
     }
+    this.lifeCycleAwares = lifeCycleAwares;
```

Comment:
For the application runner, it needs to know about the life cycle of StreamProcessors, i.e. start, shutdown and failure. So it expects the stream processor to trigger the life cycle callbacks. For example, in case of stream processor shutdown (either end of stream or use request shutdown), StreamProcessor needs to notify the application runner. So after all the stream processors finish or there is a failure, ApplicationRunner will shut down.

Btw, all these are internal life cycle methods. StreamProcessor is not exposed to the user, nor the life cycle management.